compilers/concrete-compiler/compiler/lib/Dialect/Concrete/Analysis/MemoryUsage.cpp

```diff
@@ -0,0 +1,252 @@
+        "allocation of buffer with a non-supported element-type");
+  for (auto size : shape) {
+    if (size == mlir::ShapedType::kDynamic) {
+      log_verbose() << "warning: dynamic dimension found during computation of "
+                       "memory usage. Dynamic size will be ignored";
+    }
+  }
+  auto multiply_ignore_dyn_size = [](int64_t size_1, int64_t size_2) {
+    // we don't want to multiply by a dynamic size
+    return size_2 == mlir::ShapedType::kDynamic ? size_1 : size_1 * size_2;
+  };
+
+  return elementSize * std::accumulate(shape.begin(), shape.end(), 1,
+                                       multiply_ignore_dyn_size);
+}
+
+bool isBufferDeallocated(mlir::Value buffer) {
+  for (auto user : buffer.getUsers()) {
+    if (mlir::isa<memref::DeallocOp>(user))
+      return true;
+  }
+  return false;
+}
+
+static std::optional<StringError> on_enter(memref::AllocOp &op,
+                                           MemoryUsagePass &pass) {
+
+  auto maybeBufferSize = getBufferSize(op.getResult().getType());
+  if (!maybeBufferSize) {
+    return maybeBufferSize.error();
+  }
+  // if the allocated buffer is being deallocated then count it as one.
+  // Otherwise (and there must be a problem) multiply it by the number of
+  // iterations
+  int64_t numberOfAlloc =
+      isBufferDeallocated(op.getResult()) ? 1 : pass.iterations;
+
+  auto location = locationString(op.getLoc());
+  // pass.iterations number of allocation of size: shape_1 * ... * shape_n *
+  // element_size
+  auto memoryUsage = numberOfAlloc * maybeBufferSize.value();
+
+  pass.feedback.memoryUsagePerLoc[location] += memoryUsage;
+
+  return std::nullopt;
+}
+
+static std::optional<StringError> on_enter(mlir::Operation *op,
+                                           MemoryUsagePass &pass) {
+  for (auto operand : op->getOperands()) {
+    // we only consider buffers
+    if (!mlir::isa<mlir::MemRefType>(operand.getType()))
+      continue;
+    // find the origin of the buffer
+    auto definingOp = operand.getDefiningOp();
+    mlir::Value lastVisitedBuffer = operand;
+    while (definingOp) {
+      mlir::ViewLikeOpInterface viewLikeOp =
+          mlir::dyn_cast<mlir::ViewLikeOpInterface>(definingOp);
+      if (viewLikeOp) {
+        lastVisitedBuffer = viewLikeOp.getViewSource();
+        definingOp = lastVisitedBuffer.getDefiningOp();
+      } else {
+        break;
+      }
+    }
+    // we already count allocations separately
+    if (definingOp && mlir::isa<memref::AllocOp>(definingOp) &&
+        definingOp->getLoc() == op->getLoc())
+      continue;
+
+    auto location = locationString(op->getLoc());
+
+    std::vector<mlir::Value> &visited = pass.visitedValuesPerLoc[location];
+
+    // the search would be faster if we use an unsorted_set, but we need a hash
+    // function for mlir::Value
+    if (std::count(visited.begin(), visited.end(), lastVisitedBuffer) == 0) {
+      visited.push_back(lastVisitedBuffer);
+
+      auto maybeBufferSize =
```

Comment:
I think you should compute the memoryUsagePerLoc at the end of the pass, it should be easy and avoid specific alloc op case.